src/api/client.ts

```diff
@@ -178,17 +103,49 @@ export class WsClient extends Emittery {
    * @param request - The authentication request, containing an app authentication token.
    */
   async authenticate(request: AppAuthenticationRequest): Promise<void> {
-    return this.exchange(request, (request, resolve) => {
+    this.authenticationToken = request.token;
+    return this.exchange(request, (request, resolve, reject) => {
+      const invalidTokenCloseListener = (
+        closeEvent: IsoWebSocket.CloseEvent
+      ) => {
+        reject(
+          new HolochainError(
+            "InvalidTokenError",
+            `could not connect to ${this.url} due to an invalid app authentication token - close code ${closeEvent.code}`
+          )
+        );
+      };
+      this.socket.addEventListener("close", invalidTokenCloseListener, {
+        once: true,
+      });
       const encodedMsg = encode({
         type: "authenticate",
         data: encode(request),
       });
       this.socket.send(encodedMsg);
-      // Message just needs to be sent first, no need to wait for a response or even require a flush
-      resolve(null);
+      // Wait before resolving in case authentication fails.
+      setTimeout(() => {
```

Comment:
If authentication fails, it'll still return a `resolve`. Waiting here prevents that. It's not a universal fool-proof way, ideally a ping can be sent after authenticating or Holochain returns an error or closes the websocket with a specific error code.

Reply:
Great stuff, thanks for polishing this :+1: . This idea did not occur to me.